modules/core/update/farm_update.module

```diff
@@ -0,0 +1,30 @@
+<?php
+
+/**
+ * @file
+ * Hooks and customizations for the farm_update module.
+ */
+
+/**
+ * Implements hook_rebuild().
+ */
+function farm_update_rebuild() {
+  \Drupal::service('farm.update')->rebuild();
+}
```

Comment:
This runs every time the cache is rebuilt? eg: `drush cr` ? And this was the simple way of getting this to run via `update.php`?

@mstenta I think you may have commented on this elsewhere or in chat, forgive me if that's the case!

I could see this lead to some issues during development - for example when editing & debugging the core asset/log/plan views. But maybe the solution is to just disable `farm_update` while doing this type of work?

Reply:
Ahh, from this comment: https://www.drupal.org/project/farm/issues/3220627#comment-14224994

> Alternatively, I wonder if hook_cache_flush() would work? Does that run every time that update.php is run? If so, then all we have to do is tell users to run update.php and it will update the config. But... I'm not sure if caches get cleared by update.php if there aren't any actual hook_update_N() implementations to run...

So I'm assuming `hook_cache_flush()` didn't work?

Interesting there isn't way to run [`hook_post_update_NAME()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_post_update_NAME/9.0.x) after each update 🤔

Reply:
> This runs every time the cache is rebuilt? eg: `drush cr` ?

Correct - so there are two ways to run this via Drush: `drush cr` (to rebuild all caches *and* revert config) and `drush farm_update:rebuild` (to *only* revert config).

> And this was the simple way of getting this to run via `update.php`?

Unfortunately, this still only runs via `update.php` if there are other updates to be performed. If not, then caches are not rebuilt.

So we may want to consider adding a step to the update instructions for explicitly clearing caches after updates.

> I could see this lead to some issues during development - for example when editing & debugging the core asset/log/plan views. But maybe the solution is to just disable `farm_update` while doing this type of work?

Agreed. And yea that was my thought: disable `farm_update`.

> So I'm assuming `hook_cache_flush()` didn't work?

I think `hook_cache_flush()` would also work, but the documentation for that hook states:

> Do NOT use this hook for rebuilding information. Only use it to flush custom caches.
> ...
> This hook is invoked by `drupal_flush_all_caches()`. It runs before module data is updated and before `hook_rebuild()`.

So `hook_rebuild()` seemed like the better place for this.

> Interesting there isn't way to run `hook_post_update_NAME()` after each update :thinking: 

Yea agreed! I wish there was a simple `hook_post_update()` that *always* ran after update.php or `drush updb`. That would keep our update instructions simpler, and wouldn't require an additional step for this.

Reply:
Hmm - I just found two hooks that DO fire with update.php:

- `hook_update_last_removed()`: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_update_last_removed/9.2.x
- `hook_removed_post_updates()`: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_removed_post_updates/9.2.x

Neither of these is intended for this kind of use, but in theory we could run `\Drupal::service('farm.update')->rebuild();` during one of them.

This would require some more testing IMO... and if we keep `hook_rebuild()` then it means we'd end up firing that code twice in the cases where there ARE update hooks to run (because the cache would also be cleared).